tunneltest/tunneltest.go

```diff
@@ -0,0 +1,485 @@
+		if !ok {
+			h, ok = t.Handler.(http.HandlerFunc)
+			if !ok {
+				fn, ok := t.Handler.(func(http.ResponseWriter, *http.Request))
+				if !ok {
+					return false, fmt.Errorf("invalid handler type for %q tunnel: %T", ident, t.Handler)
+				}
+
+				h = http.HandlerFunc(fn)
+			}
+
+		}
+
+		logf("serving on local %s for HTTP tunnel %q", l.Addr(), ident)
+
+		go (&http.Server{Handler: h}).Serve(l)
+
+		tt.Server.AddHost(cfg.LocalAddr, ident)
+
+		tt.mu.Lock()
+		tt.Listeners[ident] = [2]net.Listener{l, nil}
+		tt.mu.Unlock()
+
+		if err := tt.addClient(ident, cfg); err != nil {
+			return false, fmt.Errorf("error creating client for %q tunnel: %s", ident, err)
+		}
+
+		logf("registered HTTP tunnel: host=%s, ident=%s", cfg.LocalAddr, ident)
+
+	case TypeTCP:
+		h, ok := t.Handler.(func(net.Conn))
+		if !ok {
+			return false, fmt.Errorf("invalid handler type for %q tunnel: %T", ident, t.Handler)
+		}
+
+		logf("serving on local %s for TCP tunnel %q", l.Addr(), ident)
+
+		go func() {
+			for {
+				conn, err := l.Accept()
+				if err != nil {
+					log.Printf("failed accepting conn for %q tunnel: %s", ident, err)
+					return
+				}
+
+				go h(conn)
+			}
+		}()
+
+		var remote net.Listener
+
+		if t.RemoteAddrIdent != "" {
+			tt.mu.Lock()
+			remote = tt.Listeners[t.RemoteAddrIdent][1]
+			tt.mu.Unlock()
+		} else {
+			remote, err = net.Listen("tcp", t.RemoteAddr)
+			if err != nil {
+				return false, fmt.Errorf("failed to listen on %q for %q tunnel: %s", t.RemoteAddr, ident, err)
+			}
+		}
+
+		// addrIdent holds identifier of client which is going to have registered
+		// tunnel via (*tunnel.Server).AddAddr
+		addrIdent := ident
+		if t.ClientIdent != "" {
+			tt.Clients[ident] = tt.Clients[t.ClientIdent]
+			addrIdent = t.ClientIdent
+		}
+
+		tt.Server.AddAddr(remote, t.IP, addrIdent)
+
+		tt.mu.Lock()
+		tt.Listeners[ident] = [2]net.Listener{l, remote}
+		tt.mu.Unlock()
+
+		if _, ok := tt.Clients[ident]; !ok {
+			if err := tt.addClient(ident, cfg); err != nil {
+				return false, fmt.Errorf("error creating client for %q tunnel: %s", ident, err)
+			}
+		}
+
+		logf("registered TCP tunnel: listener=%s, ip=%v, ident=%s", remote.Addr(), t.IP, addrIdent)
+
+	default:
+		return false, fmt.Errorf("unknown %q tunnel type: %d", ident, t.Type)
+	}
+
+	return true, nil
+}
+
+func (tt *TunnelTest) addClient(ident string, cfg *tunnel.ClientConfig) error {
+	if _, ok := tt.Clients[ident]; ok {
+		return fmt.Errorf("tunnel %q is already being served", ident)
+	}
+
+	c, err := tunnel.NewClient(cfg)
+	if err != nil {
+		return err
+	}
+
+	go c.Start()
+	<-c.StartNotify()
+
+	tt.Clients[ident] = c
+	return nil
+}
+
+func (tt *TunnelTest) Serve(tunnels map[string]*Tunnel) error {
+	if len(tunnels) == 0 {
+		return errors.New("no tunnels to serve")
+	}
+
+	// Since one tunnels depends on others do 3 passes to start them
+	// all, each started tunnel is removed from the tunnels map.
+	// After 3 passes all of them must be started, otherwise the
```

Comment:
in my opinion removing items from a map in a `serveDeps` named  function is so implicit, can we change this naming?